src/utils/sql.ts

```diff
@@ -0,0 +1,221 @@
+
+  let output = '';
+
+  let currentDollarStringTagName = null;
+  let isString = false;
+  let isColumn = false;
+  let previousSliceEnd = 0;
+  let isSingleLineComment = false;
+  let isCommentBlock = false;
+
+  for (let i = 0; i < sqlString.length; i++) {
+    const char = sqlString[i];
+
+    if (isColumn) {
+      if (char === dialect.TICK_CHAR_RIGHT) {
+        isColumn = false;
+      }
+
+      continue;
+    }
+
+    if (isString) {
+      if (char === '\'' && !isBackslashEscaped(sqlString, i - 1)) {
+        isString = false;
+      }
+
+      continue;
+    }
+
+    if (currentDollarStringTagName !== null) {
+      if (char !== '$') {
+        continue;
+      }
+
+      const remainingString = sqlString.slice(i, sqlString.length);
+
+      const dollarStringStartMatch = remainingString.match(/^\$(?<name>[a-z_][0-9a-z_])?(\$)/i);
+      const tagName = dollarStringStartMatch?.groups?.name;
+      if (currentDollarStringTagName === tagName) {
+        currentDollarStringTagName = null;
+      }
+
+      continue;
+    }
+
+    if (isSingleLineComment) {
+      if (char === '\n') {
+        isSingleLineComment = false;
+      }
+
+      continue;
+    }
+
+    if (isCommentBlock) {
+      if (char === '*' && sqlString[i + 1] === '/') {
+        isCommentBlock = false;
+      }
+
+      continue;
+    }
+
+    if (char === dialect.TICK_CHAR_LEFT) {
+      isColumn = true;
+      continue;
+    }
+
+    if (char === '\'') {
+      isString = true;
+      continue;
+    }
+
+    if (char === '-' && sqlString.slice(i, i + 3) === '-- ') {
+      isSingleLineComment = true;
+      continue;
+    }
+
+    if (char === '/' && sqlString.slice(i, i + 2) === '/*') {
+      isCommentBlock = true;
+      continue;
+    }
+
+    // either the start of a $bind parameter, or the start of a $tag$string$tag$
+    if (char === '$') {
+      const previousChar = sqlString[i - 1];
+
+      // we are part of an identifier
+      if (/[0-9a-z_]/i.test(previousChar)) {
+        continue;
+      }
+
+      const remainingString = sqlString.slice(i, sqlString.length);
+
+      const dollarStringStartMatch = remainingString.match(/^\$(?<name>[a-z_][0-9a-z_]*)?\$/i);
+      if (dollarStringStartMatch) {
+        currentDollarStringTagName = dollarStringStartMatch.groups?.name ?? '';
+
+        continue;
+      }
+
+      continue;
+    }
+
+    if (isNamedReplacements && char === ':') {
+      const previousChar = sqlString[i - 1];
+      // we want to be conservative with what we consider to be a replacement to avoid risk of conflict with potential operators
+      // users need to add a space before the bind parameter (except after '(', ',', and '=')
+      if (previousChar !== undefined && !/[\s(,=]/.test(previousChar)) {
+        continue;
+      }
+
+      const remainingString = sqlString.slice(i, sqlString.length);
+
+      const match = remainingString.match(/^:(?<name>[a-z_][0-9a-z_]*)(?:\)|,|$|\s|::)/i);
```

Comment:
This regex does not take into account an ending `;` on the statement, so I corrected locally to `remainingString.match(/^:(?<name>[a-z_][0-9a-z_]*)(?:\)|,|\;|$|\s|::)/i);` and then also discovered it does not like when the sql statement has JSON in it such as: `UPDATE table SET value = value || '{"column::form": :formTemplateId}' where key = 'formTemplates.defaultForms'` (with replacements for `:formTemplateId` passed in)

Reply:
That's not a bug, your `:formTemplateId` is inside of a SQL string so we won't replace it because that will lead to sql injection as well as the risk of mangling data

The `;` thing may be a bug, I'd need an example of a failing query to make sure